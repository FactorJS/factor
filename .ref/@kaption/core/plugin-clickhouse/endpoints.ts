import type { EndpointMeta, EndpointResponse } from '@factor/api'
import type { Knex } from 'knex'
import type { AggregationRow, QueryParams } from '../plugin-dashboards'
import type { EventParams } from '../plugin-beacon'
import { KaptionQuery } from '../utils'
import { refineParams } from '../utils/params'

export class QueryGetClientSessions extends KaptionQuery {
  async run(
    params: {
      anonymousId: string
      projectId: string
      limit?: number
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<EventParams[]>> {
    const ch = this.kaptionClickHouse
    if (!ch)
      throw new Error('kaptionClickHouse missing')

    const { anonymousId, projectId, limit = 10 } = params

    const query = ch
      .clickhouseBaseQuerySession({ projectId })
      .where({
        anonymousId,
      })
      .limit(limit)

    const { data } = await ch.clickHouseSelect<EventParams[]>(query)

    return { status: 'success', data }
  }
}

export class QueryGetTotalSessions extends KaptionQuery {
  async run(
    params: {
      anonymousId: string
      projectId: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<number>> {
    const { anonymousId, projectId } = params
    const ch = this.kaptionClickHouse
    if (!ch)
      throw new Error('kaptionClickHouse missing')

    const query = ch
      .clickhouseBaseQuery({ projectId })
      .select(ch.client().raw('count(sessionId) as total'))
      .where({ anonymousId })
      .groupBy('anonymousId')

    const { data } = await ch.clickHouseSelect<{ total: number }[]>(query)

    const total = data[0]?.total || 0

    return { status: 'success', data: total }
  }
}

export class QueryGetDimensionList extends KaptionQuery {
  async run(params: QueryParams): Promise<EndpointResponse<AggregationRow[]>> {
    const ch = this.kaptionClickHouse
    if (!ch)
      throw new Error('kaptionClickHouse missing')
    const client = ch.client()
    const refined = refineParams(params)
    const { limit = 20, dimension, search } = refined

    if (!dimension)
      throw this.stop({ message: 'no dimension field provided' })
    const base: Knex.QueryBuilder = ch.clickhouseDateQuery({ params: refined })

    const select = [
      client.raw(`uniq(*) as count`),
      client.raw(`?? as name`, [dimension as string]),
    ]

    const dbQuery = base
      .select(select)
      .groupBy(dimension)
      .orderBy('count', 'desc')
      .whereNot({ [dimension]: '' })
      .limit(limit)

    if (search) {
      void dbQuery.whereRaw(
        client.raw(`?? like '%?%'`, [dimension as string, search]),
      )
    }

    const { data: r } = await ch.clickHouseSelect<AggregationRow[]>(dbQuery)

    return { status: 'success', data: r }
  }
}
