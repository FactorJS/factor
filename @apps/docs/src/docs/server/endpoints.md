# Endpoints and Middleware

_For operations that must happen on the server, Factor includes utilities for adding endpoints and middleware. This document discusses how to create and use them..._

## When To Use

In a normal web-app, many actions need to happen in a trusted server environment. Some example scenarios:

- Indexing or deleting data in a third-party service
- Charging a customer
- Authentication

This is because transactions handle private information and secret API keys that can't be exposed to the public.

Using endpoints and middleware allows the Factor application to securly request an action take place; and simply returns the result to the user.

## HTTP Endpoints

As endpoints are a common pattern, Factor includes a standard endpoint handling utility to make this easy.

- On the server

  - Endpoints are added via the `endpoints` filter
  - Endpoints represent a typical Factor module or class. This allows for simple resource sharing amonst similar endpoint requests.
  - The endpoint handler takes care of parsing requests and determining authentication status.

- In the client
  - The form of an endpoint request uses `$factor.$endpoint.request(args)`
  - The request method arguments:
    - `id` (required) - The ID of the endpoint to call
    - `method` (required) - The method of the endpoint class
    - `auth` (optional) - If the user must be authenticated (default: `true`)
    - `params` (optional) - Additional arguments are passed as parameters to the method (default: `{}`)

Adding an endpoint in a server accessible module/plugin:

```javascript
import { addCallback } from "@factor/tools"

addCallback("endpoints", {
  id: "myEndpoint",
  handler: { myEndpointMethod }
})

function myEndpointMethod(params) {
  return `my response with option: ${params.myOption}`
}
```

And in another plugin in the client or app:

```javascript

// Requesting an endpoint transaction from your app

import { endpointRequest } from "@factor/endpoint"

async requestEndpointMethod() {
  // Request and await transaction of endpointMethod
  const response = await endpointRequest({
    id: "myEndpoint",
    method: "endpointMethod",
    params: { myOption: 123 }
  })

  console.log(response) // "my response with option: 123"

  // Note: Authorization header automatically included with user token which is used to determine auth status
}
```

## Middleware

Endpoints are built on top of "middleware." Middleware is used as a more general utility for handling server requests to specific URLs. In Factor, middleware is easily extended and adding middleware can solve many different problems, such as generating sitemaps or handling form data.

In Node, middleware functions always take the form:

```javascript
const myMiddleware = (request, response, next) => {}
```

- _Request_ - Is the client's request to the URl
- _Response_ - Is your server's response
- _Next_ - Is a function which Node uses to determine if it should continue processing middleware at a specific URL or stop.

### Adding Middleware

To add middleware to your Factor server, all you have to do is use the `middleware` filter and a middleware object which includes the `path` and callback function(s).

As an example, middleware for creating a sitemap might look like the following:

```javascript
import { pushToFilter } from "@factor/tools"

pushToFilter("middleware", {
  path: "/sitemap.xml",
  callback: async (request, response, next) => {
    // generate sitemap
    const sitemapXML = await this.getSitemap()

    response.header("Content-Type", "application/xml")
    response.send(sitemapXML)

    // Notes: No 'next' call is needed since we don't need to continue processing other middleware
  }
})
```
