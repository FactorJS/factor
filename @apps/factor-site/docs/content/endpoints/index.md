---
title: Server Endpoints
description: Creating custom server endpoints
---

.
