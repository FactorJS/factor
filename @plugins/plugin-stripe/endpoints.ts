/* eslint-disable @typescript-eslint/no-use-before-define */
import { isNode, objectId, EndpointResponse, PrivateUser } from "@factor/api"
import { Queries as UserQueries } from "@factor/engine/user"
import { Query } from "@factor/engine/query"
import {
  FactorEndpoint,
  EndpointMeta,
  EndpointManageAction,
  EndpointMethodOptions,
} from "@factor/engine/endpoint"
import { paymentsSetting, CustomerData, stripeEnv, getStripeProducts } from "."
import Stripe from "stripe"

type RefineResult = {
  customerData?: CustomerData
  user?: PrivateUser
}
abstract class QueryPayments extends Query {
  async refine(
    params: { customerId?: string; userId?: string },
    meta: EndpointMeta,
  ): Promise<RefineResult> {
    const { customerId, userId } = params

    const out: RefineResult = {}

    if (customerId) {
      const r = await Queries.GetCustomerData.serve({ customerId }, meta)
      out.customerData = r.data
    }

    if (userId) {
      const privateDataResponse = await UserQueries.ManageUser.serve(
        {
          userId,
          _action: "getPrivate",
        },
        meta,
      )
      out.user = privateDataResponse.data
    }

    return out
  }

  async serveRequest(
    params: Parameters<this["run"]>[0],
    meta: EndpointMeta,
  ): Promise<Awaited<ReturnType<this["run"]>> & RefineResult> {
    const result = await this.serve(params, meta)

    if (result.status == "success") {
      const r = result as Awaited<ReturnType<this["run"]>> & {
        customerId?: string
        userId?: string
      }

      const { customerId, userId } = r

      const fullResponse = await this.refine({ customerId, userId }, meta)

      return { ...r, ...fullResponse }
    } else {
      return result as Awaited<ReturnType<this["run"]>> & RefineResult
    }
  }
}

export const stripeSecretKey = (): string => {
  const stripeSecretKey =
    stripeEnv() == "production"
      ? process.env.STRIPE_SECRET_KEY
      : process.env.STRIPE_SECRET_KEY_TEST

  if (!stripeSecretKey) {
    throw new Error(`stripe secret key missing: ${stripeEnv()}`)
  }

  return stripeSecretKey
}

export const getStripe = (): Stripe => {
  if (!isNode) throw new Error("Stripe is server only")

  return new Stripe(stripeSecretKey(), { apiVersion: "2020-08-27" })
}

class QueryManageCustomer extends QueryPayments {
  async run(
    params: {
      customerId?: string
      id?: string
      name?: string
      email?: string
      _action: EndpointManageAction
    },
    _meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.Customer | Stripe.DeletedCustomer> & {
      customerId: string
      customerData?: CustomerData
    }
  > {
    const stripe = getStripe()

    const { onCustomerCreated } = paymentsSetting("hooks") ?? {}

    const { _action, customerId } = params

    let customer: Stripe.Customer | Stripe.DeletedCustomer

    if (_action == "create" || (_action == "retrieve" && !customerId)) {
      const { email = "", name = "", id = "" } = params

      customer = await stripe.customers.create({
        email,
        name,
        metadata: { id },
      })

      if (onCustomerCreated) {
        await onCustomerCreated({ customer, email, id, name })
      }
    } else if (customerId) {
      customer = await stripe.customers.retrieve(customerId)
    } else {
      throw this.stop("could not get stripe customer")
    }

    return {
      status: "success",
      data: customer,
      customerId: customer.id,
    }
  }
}

class QueryPaymentMethod extends QueryPayments {
  async run(
    params: {
      customerId: string
      paymentMethodId?: string
      _action: EndpointManageAction
    },
    meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.ApiList<Stripe.PaymentMethod>> & {
      customerId: string
      customerData?: CustomerData
      setupIntent?: Stripe.SetupIntent
    }
  > {
    const { _action, customerId, paymentMethodId } = params

    if (!_action) throw this.stop({ message: "no _action provided" })
    if (!customerId) throw this.stop({ message: "no customer id" })

    const stripe = getStripe()

    // https://stripe.com/docs/api/setup_intents
    let setupIntent: Stripe.SetupIntent | undefined

    if (paymentMethodId) {
      if (_action == "create") {
        setupIntent = await stripe.setupIntents.create({
          customer: customerId,
          payment_method: paymentMethodId,
          confirm: true,
          description: "CC",
          usage: "off_session",
        })

        await this.serve(
          {
            _action: "setDefault",
            customerId,
            paymentMethodId,
          },
          meta,
        )
      } else if (_action == "setDefault") {
        if (!paymentMethodId) throw this.stop({ message: "no payment id" })
        await stripe.customers.update(customerId, {
          invoice_settings: {
            default_payment_method: paymentMethodId,
          },
        })
      } else if (_action == "delete") {
        await stripe.paymentMethods.detach(paymentMethodId)
      } else if (_action == "attach") {
        await stripe.paymentMethods.attach(paymentMethodId, {
          customer: customerId,
        })

        await this.serve(
          {
            _action: "setDefault",
            customerId,
            paymentMethodId,
          },
          meta,
        )
      }
    } else if (_action != "retrieve") {
      throw this.stop({ message: `no payment method id for ${_action}` })
    }

    const paymentMethods = await stripe.paymentMethods.list({
      customer: customerId,
      type: "card",
    })

    return { status: "success", data: paymentMethods, customerId, setupIntent }
  }
}

class QueryListSubscriptions extends QueryPayments {
  async run(
    params: {
      customerId: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.ApiList<Stripe.Subscription>>> {
    const { customerId } = params

    if (!customerId) throw this.stop({ message: "no customer id" })

    const stripe = getStripe()
    const data = await stripe.subscriptions.list({ customer: customerId })

    return { status: "success", data }
  }
}

class QueryManageSubscription extends QueryPayments {
  async run(
    params: { customerId: string } & (
      | {
          _action: "create"

          idempotencyKey: string
          priceId: string
          paymentMethodId?: string
          coupon?: string
        }
      | {
          _action: "delete"
          subscriptionId: string
          customerId: string
          note?: string
        }
      | { _action: "retrieve"; customerId?: string; subscriptionId: string }
    ),

    meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.Subscription> & {
      customerId: string
      customerData?: CustomerData
      userId?: string
      user?: PrivateUser
    }
  > {
    const { _action, customerId } = params

    if (!_action) throw this.stop({ message: "no _action provided" })

    const stripe = getStripe()

    let sub: Stripe.Subscription | undefined
    let message: string | undefined = undefined
    try {
      if (_action == "create") {
        const { customerId, paymentMethodId, priceId, coupon, idempotencyKey } =
          params

        if (!customerId) throw this.stop({ message: "no customer id" })

        // attach payment method to customer
        if (paymentMethodId && customerId) {
          await Queries.ManagePaymentMethod.serve(
            {
              customerId,
              paymentMethodId,
              _action: "attach",
            },
            meta,
          )
        }

        const { beforeCreateSubscription } = paymentsSetting("hooks") ?? {}

        let args: Stripe.SubscriptionCreateParams = {
          customer: customerId,
          items: [{ price: priceId }],
          coupon,
          expand: ["latest_invoice.payment_intent"],
          trial_period_days: 14,
        }

        if (beforeCreateSubscription) {
          args = await beforeCreateSubscription(args)
        }

        sub = await stripe.subscriptions.create(args, { idempotencyKey })

        const suffix = coupon ? ` with code ${coupon}` : ""
        message = `subscription created ${suffix}`
      } else if (_action == "retrieve") {
        const { subscriptionId } = params
        if (!subscriptionId) throw this.stop({ message: "no subscription id" })

        sub = await stripe.subscriptions.retrieve(subscriptionId)
      } else if (_action == "delete") {
        const { subscriptionId, customerId, note } = params
        if (!subscriptionId) throw this.stop({ message: "no subscription id" })

        sub = await stripe.subscriptions.del(subscriptionId)
        await stripe.customers.update(customerId, {
          metadata: { cancelReason: note ?? "" },
        })
        message = "subscription deleted"
      }
    } catch (error: unknown) {
      const e = error as Error
      throw this.stop({ message: e.message })
    }

    const { onSubscriptionUpdate } = paymentsSetting("hooks") ?? {}

    if (sub && onSubscriptionUpdate) {
      await onSubscriptionUpdate(sub)
    }

    return {
      status: "success",
      data: sub,
      customerId,
      userId: meta.bearer?.userId,
      message,
    }
  }
}

class QueryGetInvoices extends QueryPayments {
  async run(
    params: {
      customerId: string
      limit?: number
      startingAfter?: string
    },
    _meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.ApiList<Stripe.Invoice>> & {
      customerId: string
      customerData?: CustomerData
    }
  > {
    const { customerId, startingAfter, limit } = params
    const stripe = getStripe()
    const data = await stripe.invoices.list({
      customer: customerId,
      starting_after: startingAfter,
      limit,
    })

    return { status: "success", data, customerId }
  }
}

class QueryGetProduct extends QueryPayments {
  async run(
    params: {
      productId: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.Product>> {
    const { productId } = params
    const stripe = getStripe()
    const product = await stripe.products.retrieve(productId)

    return { status: "success", data: product }
  }
}

class QueryAllProducts extends QueryPayments {
  async run(
    _params: undefined,
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.Product[]>> {
    const products = getStripeProducts()

    const productIds = products
      .map((_) => _.productId)
      .filter((_) => _) as string[]

    const responsePlans = await Promise.all(
      productIds.map((productId: string) =>
        Queries.GetProduct.serve({ productId }, {}),
      ),
    )
    const data = responsePlans
      .map((product) => product.data)
      .filter((_) => _) as Stripe.Product[]

    return { status: "success", data }
  }
}

class QueryGetCoupon extends QueryPayments {
  async run(
    {
      couponCode,
    }: {
      couponCode: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.Response<Stripe.Coupon>>> {
    if (!couponCode) throw this.stop({ message: "no code was provided" })

    const stripe = getStripe()

    try {
      const data = await stripe.coupons.retrieve(couponCode)
      return { status: "success", data }
    } catch (error: unknown) {
      throw this.stop({
        message: "payment API error",
        data: error as Stripe.Errors["StripeError"],
      })
    }
  }
}

class QueryGetCustomerData extends QueryPayments {
  async run(
    {
      customerId,
    }: {
      customerId: string
    },
    meta: EndpointMeta,
  ): Promise<EndpointResponse<CustomerData>> {
    const [customer, subscriptions, invoices, paymentMethods, allProducts] =
      await Promise.all([
        Queries.ManageCustomer.serve({ customerId, _action: "retrieve" }, meta),
        Queries.ListSubscriptions.serve({ customerId }, meta),
        Queries.GetInvoices.serve({ customerId }, meta),
        Queries.ManagePaymentMethod.serve(
          { customerId, _action: "retrieve" },
          meta,
        ),
        Queries.AllProducts.serve(undefined, meta),
      ])

    const data: CustomerData = {
      subscriptions: subscriptions.data,
      customer: customer.data,
      invoices: invoices.data,
      paymentMethods: paymentMethods.data,
      allProducts: allProducts.data,
      idempotencyKey: objectId(),
    }
    return { status: "success", data }
  }
}

export const Queries = {
  ManageCustomer: new QueryManageCustomer(),
  ListSubscriptions: new QueryListSubscriptions(),
  GetInvoices: new QueryGetInvoices(),
  ManageSubscription: new QueryManageSubscription(),
  ManagePaymentMethod: new QueryPaymentMethod(),
  GetCustomerData: new QueryGetCustomerData(),
  AllProducts: new QueryAllProducts(),
  GetProduct: new QueryGetProduct(),
  GetCoupon: new QueryGetCoupon(),
}

class EndpointMethodPayments<T extends Query> extends FactorEndpoint<T> {
  constructor(options: EndpointMethodOptions<T>) {
    super({ basePath: "/payments", ...options })
  }
}

type EndpointMap = {
  [P in keyof typeof Queries]: EndpointMethodPayments<typeof Queries[P]>
}

export const getPaymentEndpointsMap = (): EndpointMap => {
  return Object.fromEntries(
    Object.entries(Queries).map(([key, query]) => {
      return [key, new EndpointMethodPayments({ key, queryHandler: query })]
    }),
  ) as EndpointMap
}

export const paymentEndpointsMap = getPaymentEndpointsMap()
export const paymentEndpoints = Object.values(paymentEndpointsMap)
