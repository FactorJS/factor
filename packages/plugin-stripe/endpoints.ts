import { objectId } from "@factor/api"
import type { EndpointResponse, PrivateUser } from "@factor/api"
import { Query } from "@factor/api/engine/query"
import { EndpointMeta, EndpointManageAction } from "@factor/api/engine/endpoint"
import Stripe from "stripe"
import { FactorUser } from "@factor/api/plugin-user"

import { CustomerData, ManageSubscriptionResult } from "./types"
import type { FactorStripe } from "."
type RefineResult = {
  customerData?: CustomerData
  user?: PrivateUser
}

abstract class QueryPayments extends Query {
  factorUser: FactorUser
  stripePlugin: FactorStripe
  constructor(settings: {
    factorUser: FactorUser
    stripePlugin: FactorStripe
  }) {
    super()

    this.factorUser = settings.factorUser
    this.stripePlugin = settings.stripePlugin
  }
  async refine(
    params: { customerId?: string; userId?: string },
    meta: EndpointMeta,
  ): Promise<RefineResult> {
    const { customerId, userId } = params

    const out: RefineResult = {}

    if (customerId) {
      const r = await this.stripePlugin.queries.GetCustomerData.serve(
        { customerId },
        meta,
      )
      out.customerData = r.data
    }

    if (userId) {
      const privateDataResponse =
        await this.factorUser.queries.ManageUser.serve(
          {
            userId,
            _action: "getPrivate",
          },
          meta,
        )
      out.user = privateDataResponse.data
    }

    return out
  }

  async serveRequest(
    params: Parameters<this["run"]>[0],
    meta: EndpointMeta,
  ): Promise<Awaited<ReturnType<this["run"]>> & RefineResult> {
    const result = await this.serve(params, meta)

    if (result.status == "success") {
      const r = result as Awaited<ReturnType<this["run"]>> & {
        customerId?: string
        userId?: string
      }

      const { customerId, userId } = r

      const fullResponse = await this.refine({ customerId, userId }, meta)

      return { ...r, ...fullResponse }
    } else {
      return result as Awaited<ReturnType<this["run"]>> & RefineResult
    }
  }
}

export class QueryManageCustomer extends QueryPayments {
  async run(
    params: {
      customerId?: string
      id?: string
      name?: string
      email?: string
      _action: EndpointManageAction
    },
    _meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.Customer | Stripe.DeletedCustomer> & {
      customerId: string
      customerData?: CustomerData
    }
  > {
    const stripe = this.stripePlugin.getServerClient()

    const { onCustomerCreated } = this.stripePlugin.setting("hooks") ?? {}

    const { _action, customerId } = params

    let customer: Stripe.Customer | Stripe.DeletedCustomer

    if (_action == "create" || (_action == "retrieve" && !customerId)) {
      const { email = "", name = "", id = "" } = params

      customer = await stripe.customers.create({
        email,
        name,
        metadata: { id },
      })

      if (onCustomerCreated) {
        await onCustomerCreated({ customer, email, id, name })
      }
    } else if (customerId) {
      customer = await stripe.customers.retrieve(customerId)
    } else {
      throw this.stop("could not get stripe customer")
    }

    return {
      status: "success",
      data: customer,
      customerId: customer.id,
    }
  }
}

export class QueryPaymentMethod extends QueryPayments {
  async run(
    params: {
      customerId: string
      paymentMethodId?: string
      _action: EndpointManageAction
    },
    meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.ApiList<Stripe.PaymentMethod>> & {
      customerId: string
      customerData?: CustomerData
      setupIntent?: Stripe.SetupIntent
    }
  > {
    const { _action, customerId, paymentMethodId } = params

    if (!_action) throw this.stop({ message: "no _action provided" })
    if (!customerId) throw this.stop({ message: "no customer id" })

    const stripe = this.stripePlugin.getServerClient()

    // https://stripe.com/docs/api/setup_intents
    let setupIntent: Stripe.SetupIntent | undefined

    if (paymentMethodId) {
      if (_action == "create") {
        setupIntent = await stripe.setupIntents.create({
          customer: customerId,
          payment_method: paymentMethodId,
          confirm: true,
          description: "CC",
          usage: "off_session",
        })

        await this.serve(
          {
            _action: "setDefault",
            customerId,
            paymentMethodId,
          },
          meta,
        )
      } else if (_action == "setDefault") {
        if (!paymentMethodId) throw this.stop({ message: "no payment id" })
        await stripe.customers.update(customerId, {
          invoice_settings: {
            default_payment_method: paymentMethodId,
          },
        })
      } else if (_action == "delete") {
        await stripe.paymentMethods.detach(paymentMethodId)
      } else if (_action == "attach") {
        await stripe.paymentMethods.attach(paymentMethodId, {
          customer: customerId,
        })

        await this.serve(
          {
            _action: "setDefault",
            customerId,
            paymentMethodId,
          },
          meta,
        )
      }
    } else if (_action != "retrieve") {
      throw this.stop({ message: `no payment method id for ${_action}` })
    }

    const paymentMethods = await stripe.paymentMethods.list({
      customer: customerId,
      type: "card",
    })

    return { status: "success", data: paymentMethods, customerId, setupIntent }
  }
}

export class QueryListSubscriptions extends QueryPayments {
  async run(
    params: {
      customerId: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.ApiList<Stripe.Subscription>>> {
    const { customerId } = params

    if (!customerId) throw this.stop({ message: "no customer id" })

    const stripe = this.stripePlugin.getServerClient()
    const data = await stripe.subscriptions.list({ customer: customerId })

    return { status: "success", data }
  }
}

export class QueryManageSubscription extends QueryPayments {
  async run(
    params: { customerId: string } & (
      | {
          _action: "create"

          idempotencyKey: string
          priceId: string
          paymentMethodId?: string
          coupon?: string
        }
      | {
          _action: "delete"
          subscriptionId: string
          customerId: string
          note?: string
        }
      | { _action: "retrieve"; customerId?: string; subscriptionId: string }
    ),

    meta: EndpointMeta,
  ): Promise<ManageSubscriptionResult> {
    const { _action, customerId } = params

    if (!_action) throw this.stop({ message: "no _action provided" })

    const stripe = this.stripePlugin.getServerClient()

    let sub: Stripe.Subscription | undefined
    let message: string | undefined = undefined
    try {
      if (_action == "create") {
        const { customerId, paymentMethodId, priceId, coupon, idempotencyKey } =
          params

        if (!customerId) throw this.stop({ message: "no customer id" })

        // attach payment method to customer
        if (paymentMethodId && customerId) {
          await this.stripePlugin.queries.ManagePaymentMethod.serve(
            {
              customerId,
              paymentMethodId,
              _action: "attach",
            },
            meta,
          )
        }

        const { beforeCreateSubscription } =
          this.stripePlugin.setting("hooks") ?? {}

        let args: Stripe.SubscriptionCreateParams = {
          customer: customerId,
          items: [{ price: priceId }],
          coupon,
          expand: ["latest_invoice.payment_intent"],
          trial_period_days: 14,
        }

        if (beforeCreateSubscription) {
          args = await beforeCreateSubscription(args)
        }

        sub = await stripe.subscriptions.create(args, { idempotencyKey })

        const suffix = coupon ? ` with code ${coupon}` : ""
        message = `subscription created ${suffix}`
      } else if (_action == "retrieve") {
        const { subscriptionId } = params
        if (!subscriptionId) throw this.stop({ message: "no subscription id" })

        sub = await stripe.subscriptions.retrieve(subscriptionId)
      } else if (_action == "delete") {
        const { subscriptionId, customerId, note } = params
        if (!subscriptionId) throw this.stop({ message: "no subscription id" })

        sub = await stripe.subscriptions.del(subscriptionId)
        await stripe.customers.update(customerId, {
          metadata: { cancelReason: note ?? "" },
        })
        message = "subscription deleted"
      }
    } catch (error: unknown) {
      const e = error as Error
      throw this.stop({ message: e.message })
    }

    const { onSubscriptionUpdate } = this.stripePlugin.setting("hooks") ?? {}

    if (sub && onSubscriptionUpdate) {
      await onSubscriptionUpdate(sub)
    }

    return {
      status: "success",
      data: sub,
      customerId,
      userId: meta.bearer?.userId,
      message,
    }
  }
}

export class QueryGetInvoices extends QueryPayments {
  async run(
    params: {
      customerId: string
      limit?: number
      startingAfter?: string
    },
    _meta: EndpointMeta,
  ): Promise<
    EndpointResponse<Stripe.ApiList<Stripe.Invoice>> & {
      customerId: string
      customerData?: CustomerData
    }
  > {
    const { customerId, startingAfter, limit } = params
    const stripe = this.stripePlugin.getServerClient()
    const data = await stripe.invoices.list({
      customer: customerId,
      starting_after: startingAfter,
      limit,
    })

    return { status: "success", data, customerId }
  }
}

export class QueryGetProduct extends QueryPayments {
  async run(
    params: {
      productId: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.Product>> {
    const { productId } = params
    const stripe = this.stripePlugin.getServerClient()
    const product = await stripe.products.retrieve(productId)

    return { status: "success", data: product }
  }
}

export class QueryAllProducts extends QueryPayments {
  async run(
    _params: undefined,
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.Product[]>> {
    const products = this.stripePlugin.getProducts()

    const productIds = products
      .map((_) => _.productId)
      .filter(Boolean) as string[]

    const responsePlans = await Promise.all(
      productIds.map((productId: string) =>
        this.stripePlugin.queries.GetProduct.serve({ productId }, {}),
      ),
    )
    const data = responsePlans
      .map((product) => product.data)
      .filter(Boolean) as Stripe.Product[]

    return { status: "success", data }
  }
}

export class QueryGetCoupon extends QueryPayments {
  async run(
    {
      couponCode,
    }: {
      couponCode: string
    },
    _meta: EndpointMeta,
  ): Promise<EndpointResponse<Stripe.Response<Stripe.Coupon>>> {
    if (!couponCode) throw this.stop({ message: "no code was provided" })

    const stripe = this.stripePlugin.getServerClient()

    try {
      const data = await stripe.coupons.retrieve(couponCode)
      return { status: "success", data }
    } catch (error: unknown) {
      throw this.stop({
        message: "payment API error",
        data: error as Stripe.Errors["StripeError"],
      })
    }
  }
}

export class QueryGetCustomerData extends QueryPayments {
  async run(
    {
      customerId,
    }: {
      customerId: string
    },
    meta: EndpointMeta,
  ): Promise<EndpointResponse<CustomerData>> {
    const [customer, subscriptions, invoices, paymentMethods, allProducts] =
      await Promise.all([
        this.stripePlugin.queries.ManageCustomer.serve(
          { customerId, _action: "retrieve" },
          meta,
        ),
        this.stripePlugin.queries.ListSubscriptions.serve({ customerId }, meta),
        this.stripePlugin.queries.GetInvoices.serve({ customerId }, meta),
        this.stripePlugin.queries.ManagePaymentMethod.serve(
          { customerId, _action: "retrieve" },
          meta,
        ),
        this.stripePlugin.queries.AllProducts.serve(undefined, meta),
      ])

    const data: CustomerData = {
      subscriptions: subscriptions.data,
      customer: customer.data,
      invoices: invoices.data,
      paymentMethods: paymentMethods.data,
      allProducts: allProducts.data,
      idempotencyKey: objectId(),
    }
    return { status: "success", data }
  }
}
