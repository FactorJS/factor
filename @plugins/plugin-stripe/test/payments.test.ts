/* eslint-disable @typescript-eslint/no-unsafe-call */
import { expect, it, describe } from "vitest"
import * as stripeEngine from "../endpoints"
import dotenv from "dotenv"
import fs from "fs"
import path from "path"
import Stripe from "stripe"
// vi.mock("../serverEmail", async () => {
//   const actual = (await vi.importActual("../serverEmail")) as Record<
//     string,
//     any
//   >
//   return {
//     ...actual,
//     getEmailSMTPService: vi.fn(() => {
//       return { sendEmail: vi.fn() }
//     }),
//   }
// })

let customer: Stripe.Customer | Stripe.DeletedCustomer | undefined
let setupIntent: Stripe.SetupIntent | undefined
// const key = Math.random().toString().slice(2, 8)

describe("stripe tests", () => {
  it("has .env file", () => {
    const dirname = new URL(".", import.meta.url).pathname
    const p = path.resolve(dirname, "../.env")
    dotenv.config({ path: path.resolve(dirname, "../.env") })

    const exists = fs.existsSync(p)

    expect(exists).toBeTruthy()
  })

  // it("create new user", async () => {
  //   const response = await userEngine.Queries.ManageUser.serve(
  //     {
  //       _action: "create",
  //       fields: { email: `arpowers+${key}@gmail.com`, fullName: "test" },
  //     },
  //     undefined,
  //   )

  //   if (!response.data) throw new Error("problem creating user")

  //   user = response.data

  //   expect(user?.userId).toBeTruthy()
  //   expect(user?.fullName).toBe("test")
  //   expect(user?.verificationCode).toBeTruthy()
  // })

  it("creates a customer for new user", async () => {
    const { status, data } = await stripeEngine.Queries.ManageCustomer.serve(
      {
        customerId: "",
        _action: "retrieve",
      },
      { server: true },
    )

    expect(status).toBe("success")
    expect(data?.id).toBeTruthy()
    customer = data
  })

  it("gets the customer", async () => {
    const { status, data } = await stripeEngine.Queries.ManageCustomer.serve(
      {
        customerId: customer?.id,
        _action: "retrieve",
      },
      { server: true },
    )

    expect(status).toBe("success")
    expect(data?.id).toBeTruthy()
  })

  it("gets the refined customer in requests", async () => {
    const { status, data, customerData, customerId } =
      await stripeEngine.Queries.ManageCustomer.serveRequest(
        {
          customerId: customer?.id,
          _action: "retrieve",
        },
        { server: true },
      )

    expect(status).toBe("success")
    expect(data?.id).toBeTruthy()
    expect(customerId).toBe(data?.id)
    expect(customerData).toBeTruthy()
    expect(Object.keys(customerData ?? {}).length).toMatchSnapshot()
  })

  it("adds a payment method", async () => {
    if (!customer?.id) throw new Error("customer required")

    const paymentMethod = await stripeEngine.getStripe().paymentMethods.create({
      type: "card",
      card: {
        number: "4242424242424242",
        exp_month: 2,
        exp_year: 2023,
        cvc: "314",
      },
    })

    const result = await stripeEngine.Queries.ManagePaymentMethod.serve(
      {
        customerId: customer?.id,
        paymentMethodId: paymentMethod.id,
        _action: "create",
      },
      { server: true },
    )

    expect(result.status).toBe("success")
    expect(result.data?.data.length).toBeGreaterThan(0)
    expect(result.setupIntent).toBeTruthy()
    setupIntent = result.setupIntent
  })

  it("gets payment methods", async () => {
    if (!customer?.id) throw new Error("customer required")

    const result = await stripeEngine.Queries.ManagePaymentMethod.serve(
      {
        customerId: customer?.id,
        _action: "retrieve",
      },
      { server: true },
    )

    expect(result.status).toBe("success")
    expect(result.data?.data.length).toBeGreaterThan(0)
  })

  it("set the default payment method", async () => {
    if (!customer?.id) throw new Error("customer required")

    const customerData = (await stripeEngine
      .getStripe()
      .customers.retrieve(customer.id)) as Stripe.Customer

    expect(customerData).toBeTruthy()
    expect(customerData.invoice_settings.default_payment_method).toBe(
      setupIntent?.payment_method,
    )
  })
})
