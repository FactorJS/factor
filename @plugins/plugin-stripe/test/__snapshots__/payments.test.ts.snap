// Vitest Snapshot v1

exports[`stripe tests > gets the refined customer in requests 1`] = `6`;
